packages/@react-aria/utils/src/useId.ts

```diff
@@ -44,3 +44,15 @@ export function mergeIds(a: string, b: string): string {
 
   return b;
 }
+
+export function useSlotId(): string {
+  let [id, setId] = useState(useId());
+  useLayoutEffect(() => {
```

Comment:
I think this is one of those times where this is the more appropriate of the two. We can change the id back to null if it isn't in the dom before the repaint. I don't know how it affects voice over though, this may still be an unnecessary optimization.